mountpoint-s3/src/data_cache/express_data_cache.rs

```diff
@@ -1,18 +1,135 @@
 use crate::object::ObjectId;
+use std::collections::HashMap;
 
 use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheError, DataCacheResult};
 
 use async_trait::async_trait;
+use base64ct::{Base64, Encoding};
 use bytes::BytesMut;
 use futures::{pin_mut, StreamExt};
 use mountpoint_s3_client::error::{GetObjectError, ObjectClientError};
-use mountpoint_s3_client::types::{GetObjectParams, GetObjectRequest, PutObjectParams};
-use mountpoint_s3_client::{ObjectClient, PutObjectRequest};
+use mountpoint_s3_client::types::{
+    ChecksumMode, GetObjectParams, GetObjectRequest, PutObjectSingleParams, UploadChecksum,
+};
+use mountpoint_s3_client::ObjectClient;
+use mountpoint_s3_crt::checksums::crc32c::{self, Crc32c};
 use sha2::{Digest, Sha256};
 use tracing::Instrument;
 
+use mountpoint_s3_client::checksums::crc32c_from_base64;
+#[cfg(test)]
+use proptest_derive::Arbitrary;
+
 const CACHE_VERSION: &str = "V1";
 
+/// Metadata about the cached object to ensure that the object we've retrieved is the one we were
+/// wanting to get (and avoid collisions with the key).
+/// On miss, bypass the cache and go to the main data source.
+#[cfg_attr(test, derive(Arbitrary))]
+#[derive(Debug, PartialEq, Eq)]
+struct BlockHeader {
+    version: String,
+    block_idx: BlockIndex,
+    block_offset: u64,
+    etag: String,
+    source_key: String,
+    source_bucket_name: String,
+    header_checksum: u32,
+}
+
+impl BlockHeader {
+    pub fn new(block_idx: BlockIndex, block_offset: u64, cache_key: &ObjectId, source_bucket_name: &str) -> Self {
+        let header_checksum = Self::get_header_checksum(block_idx, block_offset, cache_key, source_bucket_name).value();
+        Self {
+            version: CACHE_VERSION.to_string(),
+            block_idx,
+            block_offset,
+            etag: cache_key.etag().as_str().to_string(),
+            source_key: cache_key.key().to_string(),
+            source_bucket_name: source_bucket_name.to_string(),
+            header_checksum,
+        }
+    }
+
+    /// Get the S3 key this block should be written to or read from.
+    pub fn to_s3_key(&self, prefix: &str) -> String {
+        let hashed_cache_key = hex::encode(
+            Sha256::new()
+                .chain_update(&self.source_key)
+                .chain_update(&self.etag)
+                .finalize(),
+        );
+        format!("{}/{}/{:010}", prefix, hashed_cache_key, self.block_idx)
+    }
+
+    /// Convert to object metadata that is HTTP header safe (ASCII only)
+    pub fn to_headers(&self) -> HashMap<String, String> {
+        let source_key_encoded = Base64::encode_string(self.source_key.as_bytes());
+        HashMap::from([
```

Comment:
Is it a reasonable risk that headers may start going over `2KiB`? E.g. if S3 allows bigger keys in future.

We could check that the total size of the headers (keys + "x-amz-meta" + values) is less than 2KIB and do not do the request if it is not.